frontends/p4/typeChecking/typeChecker.cpp

```diff
@@ -960,6 +960,8 @@ const IR::Node* TypeInference::preorder(IR::Declaration_Instance* decl) {
             prune();
             return decl;
         }
+        auto *learn = clone();
+        (void)type->apply(*learn);
```

Comment:
This applying a clone and ignoring the return value just to update the typeMap still bugs me -- its a symptom of the messiness of having the typeMap separate from the type field(s) in the IR that I really would like to clean up eventually.

Reply:
Removing the referenceMap should be easier.
The typechecking pass has grown quite complicated. Some of the complications are probably unnecessary - like maintaining canonical types. But I don't see how you can maintain such a complicated graph of types without an auxiliary data structure. 
The most annoying thing about type checking is that it needs to run over and over, any time the graph changes. But I think that the alternative of having each pass compute types on its own is worse. At least this way the type checking is centralized in a single pass.